[Rules] Add operators to convert integer values

TD-er opened the pull request:
For example:

- toBin
- toHex
- bitRead
- bitSet
- bitClear
- bitWrite
- XOR
- AND
- OR
- div
- mod

Fixes: #3444

TD-er committed:
[Rules] Add operators to convert integer values 

For example:

- toBin
- toHex
- bitRead
- bitSet
- bitClear
- bitWrite
- XOR
- AND
- OR
- div
- mod

TD-er committed:
[Docs] Document the effects of external hosted CSS/JS files

tonhuisman commented:
There was a request for `abs` in the forum recently, could be nice to have that too 😉

TD-er commented:
Made a test build:

Edit:
See below for newer build

@tonhuisman Sure, will add `abs` too

TD-er committed:
Add ABS function

TD-er commented:
@tonhuisman Done.

TD-er committed:
[Rules] Add trigonometric functions

cos, sin, tan etc. and their inverse functions.
Also possible to add a suffix `_d` to convert the angles to or from degrees.

TD-er commented:
Added trigonometric functions and also deg/rad conversions to make it easier to use with degrees.

TD-er committed:
[Rules] Use bitwise operators only on unsigned ints

TD-er commented:
OK, the bitwise operators were not working as they should, so it has now been fixed.

Bitwise operator functions now only work on unsigned ints.
It is very likely bitwise operators will be using 32 bits, or maybe even slightly more as some may be stripped out.
Therefore the bitwise functions can handle upto 64 bit values.
Those still cannot be used on event values, but better be prepared :)

Also the trigonometric functions added quite some extra to the binary files, so those are now made optional (and thus excluded on build size limited builds).

TD-er committed:
Fix build warning

TD-er commented:
A testbuild:

- [ESPEasy_ESP82xx_mega-20201227-7-PR_3439.zip](https://www.dropbox.com/s/ghekf55fj68fxto/ESPEasy_ESP82xx_mega-20201227-7-PR_3439.zip?dl=0)
- [ESPEasy_ESP32_mega-20201227-7-PR_3439.zip](https://www.dropbox.com/s/9xkd51j1g9yje7x/ESPEasy_ESP32_mega-20201227-7-PR_3439.zip?dl=0)

TD-er committed:
[Rules] Change order parameters bitwrite

TD-er committed:
[Rules] Increase accuracy in Calculate and variables from float to double

This will help people when trying to store 32 bit ints in variables.

TD-er committed:
Include P090 CCS811 in the builds again

TD-er committed:
Increase accuracy in rules computation by using more decimals

TD-er committed:
Merge branch 'mega' into feature/bitwise_operations

TD-er committed:
[Rules] Improve calculation resolution and compares.

Using proper floating point compare functions.

TD-er committed:
[Build] Reduce build size of OTA builds including new rules operators

TD-er commented:
New test build:

- [ESPEasy_ESP82xx_mega-20201227-32-PR_3439.zip](https://www.dropbox.com/s/j3jbmrr1svngfi9/ESPEasy_ESP82xx_mega-20201227-32-PR_3439.zip?dl=0)
- [ESPEasy_ESP32_mega-20201227-32-PR_3439.zip](https://www.dropbox.com/s/hynvf8dkckibmg2/ESPEasy_ESP32_mega-20201227-32-PR_3439.zip?dl=0)

Main improvements:
- Using proper checks for floating point compares
- Fixing the issues with the time compare mentioned here: https://github.com/letscontrolit/ESPEasy/pull/3131#issuecomment-649824813
- Variables are now stored as `double` instead of `float`, so 32 bit ints can now be stored in a variable without loosing resolution.
- Computations are now done using more decimals in rules.
- `[var#1]` will now output much more decimals but trailing 0's will be stripped (unless nr of decimals is set using formatting)
- `%v1%` will now output in 6 decimals but trailing 0's will be removed
- since `%v1%` is using 6 decimals while the floating point values are stored with a much higher resolution, the converted values may experience less errors. See example below for illustration.

Illustration rules for testing the improved resolution and predictability of comparing floats:

```
 on eventname do
   let,4,=1/3
   let,4,=[var#4]*3
   LogEntry,'Test Var: [var#4]  pct_v: %v4%'
   if 1 > [var#4]
     LogEntry,'1 > [var#4]'
   else 
     LogEntry,'1 =< [var#4]'
   endif
 endon
```

Log output:
```
7247376 : Info   : EVENT: eventname=1,2,3,4
7247384 : Info   : ACT  : let,4,=1/3
7247391 : Info   : ACT  : let,4,=0.33333333333333*3
7247402 : Info   : ACT  : LogEntry,'Test Var: 0.99999999999999  pct_v: 1'
7247407 : Info   : Test Var: 0.99999999999999  pct_v: 1
7247414 : Debug  : conditionMatch: _1 > 0.99999999999999_ val1: 1.00 "1" val2: 1.00 "0.99999999999999"  = false
7247425 : Info   : ACT  : LogEntry,'1 =< 0.99999999999999'
7247431 : Info   : 1 =< 0.99999999999999
```

TD-er committed:
[Build] Move PMSx003 to 'normal' builds

TD-er committed:
[Cleanup] Wrap <div> markup in a separate function

To reduce build size and make code better readable

TD-er committed:
[Cleanup] Reduce build size by replacing addHtml(String(....)) to addHtmlInt

tonhuisman commented on src/src/WebServer/Markup_Forms.cpp:
```diff
@@ -32,13 +32,7 @@ void addFormSeparator(int clspan)
-  String html;
-
-  html.reserve(40 + text.length());
-  html += F("<div class='note'>Note: ");
```
Is that _Note:_ prefix no longer part of a note? I find it quite useful.

TD-er replied:
Was not my intention to remove it.
This is the reason why I love it when others also have a look at the code 👍

TD-er committed:
Restore accidentally removed "Note:" prefix for notes

uzi18 commented:
@TD-er are they only strings functions? 
if so user should have some (3-4) private user strings to operate on.

On the other hand we should have some functions for operations like:
```
   let,3,32
   let,4,-1
   let,5,=abs([var#4])*bitRead([var#4],5)

or
IF bitRead([var#4],6) THEN ...
```
Sure it is more difficult but more usable.

TD-er commented:
Right now a command does not return a string which can be used to replace parts in the rules parsing.
For this to work, you need to add a new type of internal commands which can be called immediately during rules parsing and then return a string.
Or we must extend the Calculate function to support longer operators (`abs` is then an operator).
This will complicate the parsing of the Calculate function quite a bit more I am afraid.

For example we now use braces as a scope indicator for things that need to be parsed first, but we don't have a mechanism to check if the braces are preceded with an operator.

So I do understand why your suggestion may be a very good one to have, but I'm afraid it will make the Calculate parse much more complex and I'm really not that good in writing grammar parsers (that was one of the courses at university I had to take several times to pass and I still don't like it)

TD-er commented:
I just took a quick look at the Calculate code.
It is now using a 1 character token parser to check if something is an operator or a numerical. (or an unary operator like the `!`)
Maybe we can do some pre-parsing to match commands like those and change them to add a special character in the line to mark special state, just like the unary operator.
But still the data within the braces cannot be "solved" without knowing it is a special operator as you may have a number of parameters in there.

Another approach is to convert your suggested format into the curly braces notation added in this PR.

The other thing you suggested is this one:
```IF bitRead([var#4],6) THEN ...```

That's a tricky one as we now require an operator in if statements.
If there is none, you explicitly mean `<>0` and I don't know what else may get broken by assuming this.

tonhuisman commented:
Maybe there is a way to handle this, as currently the `IF` doesn't require a `THEN`, we could add full expression resolving when the `THEN` is used. The current behavior could then stay the same (to keep performance and backward compatibility). But it could still be quite an undertaking...
I'll do some experiments, shortly

NB: It has been requested/expected before (where/as the `IF` doesn't support expressions).

TD-er commented:
if does support expressions.

```
if {abs:%eventvalue%} > 2 then
```
I see no reason why this should not work (not tested it)

tonhuisman commented:
I seem to recall it doesn't support stuff like `if [task#var] = %v1% / 2 ...` where the expression `%v1% / 2` isn't calculated? It surprised me at the time, but the example code provided then wasn't working, where using an intermediate like `let,2,%v1%/2` and `if [task#var] = %v2%` did work as expected. Not sure if I can find that issue back in the huge stack of stuff... Ah, found it: #3121 but it's about the `%` (modulus) operation, which shouldn't matter I guess.

uzi18 commented:
I just introduced the idea to have something more human readable (produce less bugs) and more effective scripts

TD-er commented:
> I just introduced the idea to have something more human readable (produce less bugs) and more effective scripts

Yep and I do agree it is way better readable.
If it appears to be not too hard to implement, I will undo all the (math related) implemented replacement things I made here to do it, as it indeed is way more readable.
But so far I have not managed to think of a way which will not need a lot of work on the calculate parser.
And that's a part I would rather not mess with a lot as I'm not that good with writing grammar parsers.

TD-er committed:
[Rules] Parse hex numbers when prefixed with 0x (uint only)

Currently only parsing hex numbers for the "bit-xxxx" functions.
Those only accept uint numerical values, which can either be in base 10 or base 16 (with "0x" prefix)

TD-er committed:
[Cleanup] Remove String construction with = "";

A String is default constructed as an empty String, so no need to add the extra step of constructing it with an empty char array.

TD-er committed:
[Rules] Allow HEX values in formula, rules, etc.

Only place where the HEX values are not yet supported is in the web interface where int/float values are entered.

TD-er commented:
Made a new test build:
- [ESPEasy_ESP82xx_mega-20201227-39-PR_3439.zip](https://www.dropbox.com/s/snwl5dgjpf2rzdz/ESPEasy_ESP82xx_mega-20201227-39-PR_3439.zip?dl=0)
- [ESPEasy_ESP32_mega-20201227-39-PR_3439.zip](https://www.dropbox.com/s/a0cdblacgyp9g27/ESPEasy_ESP32_mega-20201227-39-PR_3439.zip?dl=0)

Main new feature is to allow the use of "HEX" numbers in formulas, rules, etc.
A hex number needs to have `0x` as prefix.

For example an event with a HEX and float value as event values:
```
Event,eventname2=0xa5,12.3
```

Rules test:
```
on eventname2 do
  let,1,%eventvalue1%*%eventvalue2%
  LogEntry,'cos([var#1]) = {cos_d:[var#1]} Inverse: {acos_d:{cos_d:[var#1]}}'
  LogEntry,'bitset(%eventvalue1%) = {bitset:0:%eventvalue1%} {tobin:{bitset:0:%eventvalue1%}}'
  LogEntry,'bitclear(%eventvalue1%) = {bitclear:0:%eventvalue1%} {tobin:{bitclear:0:%eventvalue1%}}'
  LogEntry,'bitwrite(%eventvalue1%) = {bitwrite:0:1:%eventvalue1%} {tobin:{bitwrite:0:1:%eventvalue1%}}'
endon
```

Log:
```
161723 : Info   : EVENT: eventname2=0xa5,12.3
161737 : Info   : ACT  : let,1,0xa5*12.3
161750 : Info   : ACT  : LogEntry,'cos(2029.5) = -0.64944940635652 Inverse: 2.27765646200954'
161757 : Info   : cos(2029.5) = -0.64944940635652 Inverse: 2.27765646200954
161766 : Info   : Bitwise: {bitset:0:0xa5}
161770 : Info   : Bitwise: {tobin:165}
161773 : Info   : ACT  : LogEntry,'bitset(0xa5) = 165 10100101'
161778 : Info   : bitset(0xa5) = 165 10100101
161785 : Info   : Bitwise: {bitclear:0:0xa5}
161788 : Info   : Bitwise: {tobin:164}
161791 : Info   : ACT  : LogEntry,'bitclear(0xa5) = 164 10100100'
161796 : Info   : bitclear(0xa5) = 164 10100100
161803 : Info   : Bitwise: {bitwrite:0:1:0xa5}
161807 : Info   : Bitwise: {tobin:1}
161809 : Info   : ACT  : LogEntry,'bitwrite(0xa5) = 1 1'
161815 : Info   : bitwrite(0xa5) = 1 1
```

tonhuisman commented:
I've been testing this for a bit now (😛), and all seems well.

TD-er commented:
Thanks :)
@uzi18
I have been thinking about how to implement the notation you mentioned and I will experiment a bit with it.
Maybe it is possible without changing a lot in the Calculate code.

uzi18 commented:
Thanks @TD-er, rules should be always maximally readable and  not too complex for easy things. 
If you add hex so maybe you can also bin like 0bxxxxxx or just bxxxxxx

Best wishes for all ESPeasy users and contributors for new year!

Grovkillen commented:
@uzi18 that's already on the Todo list. We just let Gijs enjoy a New Years Eve first. :grin:

TD-er commented:
> Thanks @TD-er, rules should be always maximally readable and not too complex for easy things.
> If you add hex so maybe you can also bin like 0bxxxxxx or just bxxxxxx

I will use the `0b` syntax, to make it very similar to the HEX notation.
And about the functions introduced in this PR, I think they will be either changed into the proposed notation, or I may add a translation to convert rules into this notation.

Both have their pros and cons.
But the main disadvantage of the notation used now is that it doesn't allow to use calculations as parameters.

The order in rules parsing as it is done now:
- Replace parameters like `[var#1]`, `%eventvalue1%` etc.
- Calculate.

The way I implemented it right now is that it uses the first part.
This also happens on a lot more occasions, so we already have the strange behavior that some calculations cannot be done on some occasions and thus we use the `let` function for calculations.

So I think we should loop over the 2 stages a few times to see if things were replaced.
Then converting the proposed notation to the one in this PR may still work and then we don't need to change a lot in the current parser.

TD-er committed:
[Rules] Change '!' + '=' into '<' + '>' to avoid conflict with '^'

As reported here:
https://github.com/letscontrolit/ESPEasy/pull/3131#pullrequestreview-560531284

TD-er committed:
[Rules] Disable math functions for now to make them more intuitive

TD-er committed:
[Serial] Only parse commands from serial when no task uses Serial0

To avoid people forgetting to disable serial in Tools->Advanced when having an enabled task using Serial0

TD-er committed:
Merge branch 'mega' into feature/bitwise_operations

TD-er commented:
A new test build:
- [ESPEasy_ESP82xx_mega-20201227-92-PR_3439.zip](https://www.dropbox.com/s/h6p4jz9vw7jqiup/ESPEasy_ESP82xx_mega-20201227-92-PR_3439.zip?dl=0)
- [ESPEasy_ESP32_mega-20201227-92-PR_3439.zip](https://www.dropbox.com/s/xv29oux4qqy2yw4/ESPEasy_ESP32_mega-20201227-92-PR_3439.zip?dl=0)

I removed the cos/sin/tan and div/mod functions (mod was already present and div I must think about for integers) so we can make them more intuitive.
Also changed the handling of Serial port, so it no longer needs to be disabled when you use a task using Serial0.

TD-er committed:
Merge branch 'mega' into feature/bitwise_operations

TD-er committed:
Fix build  issue after merging other PRs

TD-er committed:
[Rules] Add support for binary numbers with 0b prefix

TD-er committed:
[Rules] Remove limit of nr. of variables

The index of an internal variable can now be any value that fits an `uint32_t`.

tonhuisman commented on src/src/Helpers/SystemVariables.cpp:
```diff
@@ -196,13 +198,13 @@ void SystemVariables::parseSystemVariables(String& s, boolean useURLencode)
+    if (validUIntFromString(s.substring(v_index + 2), i)) {
+      String key = "%v" + String(i) + '%';
       if (s.indexOf(key) != -1) {
         const bool trimTralingZeros = true;
```
That one reads as `trimTralala zeros` 😛 suggestion: `trimTrailingZeros` (not sure if I saw this one earlier or it's also somewhere else, but I'm sure I saw it before)

TD-er replied:
Ah had to read it twice to know what you meant.
And it is copy/paste, so I will fix it in both places :)

TD-er committed:
[Rules] Improve readability of Rules Calculate function

Return code to `enum class` and log each call to calculate if it returns an error.

Also act on the return code of the `Calculate` function where used.

TD-er committed:
[Rules] Wrap rules calculation functions in a struct + cleanup

Some globally declared variables (`sp`) and functions (`push`, `pop`) were just too generic they could lead to issues in the future.
By wrapping them in a struct, it is more clear what is happening and what should be used to access the calculation functions.

TD-er committed:
[Rules] Fix out of bound crash with calculate ending with '=' + cleanup

Added some defines to improve readability.
Moved elaborate check to separate `is_number` function.

TD-er committed:
[Rules] Add preliminary support for math functions like sqrt, sin, etc.

Not yet complete

TD-er committed:
[Cleanup] Change typo trimTrailingZeros

As reported here: https://github.com/letscontrolit/ESPEasy/pull/3439#discussion_r553986899

TD-er committed:
[Rules] Add missing functions (sin/cos etc)

TD-er committed:
[Rules] Fix typo on limited size builds

TD-er commented:
Added the suggested format as suggested by @uzi18 
So would be nice if this could be tested.

A test build is running, but most of you following this PR are capable of building your own :)

TD-er commented:
Edit: For new test build, see below

TD-er commented:
Hmm already thought of a bug that's still in there.
I need to check for `asin(` first before checking `sin(`
So the arc-versions of those functions will very likely not yet work.

TD-er committed:
Merge branch 'mega' into feature/bitwise_operations

TD-er committed:
[Rules] Fix order of parsing "arc" functions

TD-er commented:
OK, added a fix for it, not yet included in the test build though.

tonhuisman commented:
Found something different, but slightly related, on the {} string functions:
Rules:
```
on Clock#Time=All,**:** do
  LogEntry,"Clock:%eventvalue%,%eventvalue2%"
  Let,3,{substring:3:4:"%eventvalue2%"}
  Let,2,{substring:0:2:"%eventvalue2%"}
  Let,1,=sin(90)
  TaskValueSet,12,1,[Var#1]
  TaskValueSet,12,2,=cos(%v2%)
  TaskValueSet,12,3,=tan(%v3%)
endon
```
Initially I didn't have the quotes around %eventvalue2%, but both cases don't work (that's why I left it in), probably the colon in the time is seen as an argument separator for the {} string functions.
Result:
```
1360445 : Info   : ACT  : LogEntry,"Clock:Sat,14:25"
1360447 : Info   : Clock:Sat,14:25
1360450 : Info   : Bitwise: {substring:3:4:"14}
1360452 : Info   : ACT  : Let,3,{substring:3:4:"14:25"}
1360454 : Info   : Bitwise: {substring:3:4:"14}
1360456 : Error  : Calculate: Unknown token input: {substring:3:4:"14:25"} = 0
1360460 : Info   : Bitwise: {substring:0:2:"14}
1360461 : Info   : ACT  : Let,2,"1
1360466 : Info   : ACT  : Let,1,=sin(90)
1360475 : Info   : ACT  : TaskValueSet,12,1,0.89399666360056
1360483 : Info   : ACT  : TaskValueSet,12,2,=cos(1)
1360491 : Info   : ACT  : TaskValueSet,12,3,=tan(0)
```
My workaround to get the minutes from the time is this:
```
  Let,4,{substring:0:%eventvalue2%                                                           } // ~60 spaces added
```
(Will cause issues between 00:00..01:59 though)
Result:
```
717518 : Info   : ACT  : LogEntry,"Clock:Sat,14:47"
717520 : Info   : Clock:Sat,14:47
717525 : Info   : Bitwise: {substring:0:14:47                                                             }
717527 : Info   : ACT  : Let,4,47
```
So to sum up the issues:
- String functions don't handle strings containing colons well
- String functions don't handle quoted strings (to 'hide' any colons)

Maybe the arguments should be parsed using `parseString()` ?

TD-er commented:
Hmm looks strange indeed.
Have to check to see what's the problem here.

Also very strange to see that adding spaces may "fix" it.

It may have something to do with the order in which variables and these string operations are being parsed.
Right now they are not parsed using any preferred order and using a stack, like the "calculate" function does.
The parsing of variables and these {} wrapped commands is only done once.
Maybe it can indeed be fixed very easy by using the `parseString` call, but I don't know what the side effects may be.

TD-er commented:
Hmm indeed the parsing in `get_next_argument` isn't that intelligent and should use the same parsing the arguments of commands.

tonhuisman commented:
> Also very strange to see that adding spaces may "fix" it.

I added them to avoid potential 'index out of range' errors, not needed if range checking is done properly.

uzi18 commented:
But you have got %syshour% and %sysmin% available

TD-er commented:
> But you have got %syshour% and %sysmin% available

Sure but it is indicative of a problem with the parsing.

TD-er committed:
[Rules] use parseString on string operation commands

TD-er commented:
@tonhuisman I fixed it.
Please try to test it :)

TD-er commented:
See below for latest test build

tonhuisman commented:
I've tried the above testbuild 113, and it does seem to process the {} string functions, but while parsing my calculations, it complains a bit:
```
552240 : Info   : EVENT: Clock#Time=Sun,21:21
552271 : Info   : ACT  : LogEntry,"Clock:Sun,21:21"
552273 : Info   : Clock:Sun,21:21
552276 : Info   : Bitwise: {substring:0:21:21}
552278 : Info   : ACT  : Let,4,21
552282 : Info   : Bitwise: {substring:3:4:"21:21"}
552283 : Info   : ACT  : Let,3,2
552286 : Info   : Bitwise: {substring:0:2:21}
552287 : Info   : ACT  : Let,2,21
552290 : Info   : ACT  : Let,1,=sin(90)
552292 : Error  : Calculate: Unknown token
552293 : Error  : Calculate: Unknown token
552296 : Info   : ACT  : TaskValueSet,12,1,0
552302 : Info   : ACT  : TaskValueSet,12,2,=cos(21)
552304 : Error  : Calculate: Unknown token
552306 : Error  : Calculate: Unknown token
552309 : Info   : ACT  : TaskValueSet,12,3,=tan(2)
552311 : Error  : Calculate: Unknown token
552313 : Error  : Calculate: Unknown token
552317 : Info   : ACT  : TaskValueSet,12,4,=tan(21)
552319 : Error  : Calculate: Unknown token
552320 : Error  : Calculate: Unknown token
```
But maybe I missed a note on a different way of using these functions? (I did try without the `=` in front)

TD-er commented:
I will check why the `=` sign may cause "unknown token" errors.
By the way, if you try to call sin/cos etc. with degrees, please use the `cos_d` function, otherwise it may look as it is not working as expected, as the default is in radians.

TD-er commented:
Were you running with the latest code?
```
5061827 : Info   : EVENT: eventname2=1,2,3,4
5061842 : Info   : ACT  : let,1,1*2
5061849 : Info   : ACT  : let,2,sqrt(2)
5061863 : Info   : ACT  : let,3,=log(1)
5061872 : Info   : ACT  : let,4,ln(1)
5061879 : Info   : ACT  : let,5,=sin(1)
5061888 : Info   : ACT  : let,6,cos(1)
5061895 : Info   : ACT  : let,7,cos_d(1)
5061905 : Info   : Bitwise: {substring:0:3:1}
5061908 : Info   : ACT  : logentry,1
5061911 : Info   : 1
5061918 : Info   : Bitwise: {substring:2:4:3}
5061921 : Info   : ACT  : logentry,{substring:2:4:"3"}
5061925 : Info   : Bitwise: {substring:2:4:3}
5061929 : Info   : {substring:2:4:"3"}
5061934 : Info   : ACT  : LogEntry,'sqrt = 1.4142135623731'
5061938 : Info   : sqrt = 1.4142135623731
```

With these rules:
```
on eventname2 do
  let,1,%eventvalue1%*%eventvalue2%
  let,2,sqrt([var#1])
  let,3,=log(%eventvalue1%)
  let,4,ln(%eventvalue1%)
  let,5,=sin(%eventvalue1%)
  let,6,cos(%eventvalue1%)
  let,7,cos_d(%eventvalue1%)
  logentry,{substring:0:3:%eventvalue%}
  logentry,{substring:2:4:"%eventvalue3%"}
  LogEntry,'sqrt = [var#2]'
endon
```

TD-er committed:
[Rules] Add some basic math functions + documentation

Moved `abs` to also be the more intuitive `abs(x)` notation.
Added:
* Exp
* Sq
* Round

TD-er commented:
See below for newer build.

TD-er committed:
[MQTT] Fix status update over MQTT for internal commands

As discussed here: https://github.com/letscontrolit/ESPEasy/issues/3444#issuecomment-757894274

TD-er committed:
[MQTT] Also strip %valname% in topic if task index is invalid

tonhuisman commented:
> Were you running with the latest code?

Yes I was, but by default I use a `test` build, and apparently this calculation stuff is excluded in the memory-tight builds, so that's why it didn't work 😊 
Now that I've tested with a `normal` build, all is working as expected 👍

TD-er committed:
[MQTT] Small optimizations in parsing publish topics

TD-er committed:
[Rules] Add option to each task to combine all values in a single event

TD-er commented:
Just added some new sweetness :)

![image](https://user-images.githubusercontent.com/3751318/104244316-761be580-5462-11eb-9511-e701dfe00dff.png)

TD-er commented:
See for latest build below
Now with fixes for #3444 and the issues discussed at [the forum](https://www.letscontrolit.com/forum/viewtopic.php?p=49470#p49470)

TD-er committed:
[MQTT] Fix remove of '//' in topic

Problem occurred when `///` -> `/` was expected.
Need to replace it in a while loop.

GravityRZ commented:
> New test build:
> 
> * [ESPEasy_ESP82xx_mega-20201227-118-PR_3439.zip](https://www.dropbox.com/s/12m3ju7gso7wb1e/ESPEasy_ESP82xx_mega-20201227-118-PR_3439.zip?dl=0)
> * [ESPEasy_ESP32_mega-20201227-118-PR_3439.zip](https://www.dropbox.com/s/ldq1g0ap8rrowlu/ESPEasy_ESP32_mega-20201227-118-PR_3439.zip?dl=0)
> 
> Now with fixes for #3444 and the issues discussed at [the forum](https://www.letscontrolit.com/forum/viewtopic.php?p=49470#p49470)

multiple %eventvalue% functionality works
Maybe put a reminder that if you select this you MUST USE #All to trigger events
Triggers will not work if you use individual variables like #Count or #Time

TD-er commented:
That's a good one, I will add it in the note also.
It is mentioned in the documentation.

TD-er committed:
[Rules] Add hint about changed event when combining all event values

TD-er committed:
[Cleanup] Uncrustify changed files of this PR

TD-er committed:
[Cleanup] Fix issues compile caused by uncrustify

TD-er committed:
[Minimal OTA] Make minimal OTA build fit again

TD-er committed:
Fix build issue on ESP32

TD-er commented:
See below for new test build.


Main change is some more trickery to make the minimal OTA builds fit again.

Let's hope I did not break anything in the process.

TD-er committed:
Accidentally removed C013 from the minimal OTA build

tonhuisman commented:
It looks quite OK, I think, though I have 1 remark.
I've defined a Dummy device, configured as Quad, and all values are set to 6 decimals. I've enabled the new setting for a combined event (mostly to reduce the log output 😄), and now I again see that the internal  logging of the Dummy device outputs its value with a fixed 2 decimals output:
```
251198 : Info   : Dummy: value 1: 1.00
251199 : Info   : Dummy: value 2: 2.00
251199 : Info   : Dummy: value 3: 3.00
251200 : Info   : Dummy: value 4: 4.00
251210 : Info   : EVENT: Dummy#All=1.000000,2.000000,3.000000,4.000000
```
I would kind of expect it to do the logging with the configured number of decimals, just like the value is passed to the event?
But maybe that is an issue with many more plugins...

tonhuisman commented on src/src/ESPEasyCore/ESPEasyRules.cpp:
```diff
@@ -1384,29 +1384,28 @@ void createRuleEvents(struct EventStruct *event) {
+  if (Settings.CombineTaskValues_SingleEvent(event->TaskIndex)) {
     String eventString;
-    eventString.reserve(32); // Enough for most use cases, prevent lots of memory allocations.
+    eventString.reserve(128); // Enough for most use cases, prevent lots of memory allocations.
```
To save a few bytes of memory, reserving 64 bytes would probably be enough for most use-cases (15 character taskname + `#All=` + 4x 10 digits + 3x `,`)

TD-er replied:
Well the fields for task name can be 25 or 40 bytes long (not sure) plus the values, so that's "quite a lot".
I could set it to 64, but it is not going to matter a lot as the String itself is almost immediately after its creation copied to the event queue.
N.B. Copied, not moved.
So after the copy, the new String has the exact size it needs.

Also I do like to have those reserve sizes a power of 2, or at least a multiple of some power of 2 (not too small, so not '1' ;) ) to limit heap fragmentation.

The idea behind it was that allocating too much isn't a big deal (here) but allocating too little is.
So changing to 64 bytes is almost always fine too, as it will probably fit like 99% of the time.

Grovkillen commented:
I setup a Dummy with four values and expected this rule to work:

```
On DummyTest#All Do
  LogEntry,[DummyTest#test1]:[DummyTest#test2]:[DummyTest#test3]:[DummyTest#test4]
EndOn
```

Nether does this one: 
```
On DummyTest#All Do
  logentry,%eventvalue1%:%eventvalue2%:%eventvalue3%:%eventvalue4%
EndOn
```

This one gives an "too many arguments" error:
```
On DummyTest#All Do
  logentry,%eventvalue1% %eventvalue2% %eventvalue3% %eventvalue4%
EndOn
```
This one works but output is removing trailing zeros.. I guess that's ok and by design:
```
On DummyTest#All Do
  logentry,%eventvalue1%
  logentry,%eventvalue2%
  logentry,%eventvalue3%
  logentry,%eventvalue4%
EndOn
```

```
430818: EVENT: DummyTest#All=0,0.00,0.0000,0.000000
430833: ACT : logentry,0
430835: 0
430837: ACT : logentry,0.00
430839: 0
430842: ACT : logentry,0.0000
430845: 0
430847: ACT : logentry,0.000000
430849: 0
```

I'm gonna test the operators now.

TD-er commented:
> I setup a Dummy with four values and expected this rule to work:
> 
> ```
> On DummyTest#All Do
>   LogEntry,[DummyTest#test1]:[DummyTest#test2]:[DummyTest#test3]:[DummyTest#test4]
> EndOn
> ```
> 
> Nether does this one:
> 
> ```
> On DummyTest#All Do
>   logentry,%eventvalue1%:%eventvalue2%:%eventvalue3%:%eventvalue4%
> EndOn
> ```
What happens if you use quotes?
This is a strange one and I guess it tries to match it as a time? 
But anyway, I don't see why these two are not working.

> This one gives an "too many arguments" error:
> 
> ```
> On DummyTest#All Do
>   logentry,%eventvalue1% %eventvalue2% %eventvalue3% %eventvalue4%
> EndOn
> ```

That's correct behavior, as a space is also a separator, so use quotes on the last parameter. 




> This one works but output is removing trailing zeros.. I guess that's ok and by design:
> 
> ```
> On DummyTest#All Do
>   logentry,%eventvalue1%
>   logentry,%eventvalue2%
>   logentry,%eventvalue3%
>   logentry,%eventvalue4%
> EndOn
> ```
> 
> ```
> 430818: EVENT: DummyTest#All=0,0.00,0.0000,0.000000
> 430833: ACT : logentry,0
> 430835: 0
> 430837: ACT : logentry,0.00
> 430839: 0
> 430842: ACT : logentry,0.0000
> 430845: 0
> 430847: ACT : logentry,0.000000
> 430849: 0
> ```

If you need spaces, use quotes :)

TD-er commented:
> It looks quite OK, I think, though I have 1 remark.
> I've defined a Dummy device, configured as Quad, and all values are set to 6 decimals. I've enabled the new setting for a combined event (mostly to reduce the log output 😄), and now I again see that the internal logging of the Dummy device outputs its value with a fixed 2 decimals output:
> 
> ```
> 251198 : Info   : Dummy: value 1: 1.00
> 251199 : Info   : Dummy: value 2: 2.00
> 251199 : Info   : Dummy: value 3: 3.00
> 251200 : Info   : Dummy: value 4: 4.00
> 251210 : Info   : EVENT: Dummy#All=1.000000,2.000000,3.000000,4.000000
> ```
> 
> I would kind of expect it to do the logging with the configured number of decimals, just like the value is passed to the event?
> But maybe that is an issue with many more plugins...

Yep, that was probably an old bug... I will have a look at it.

Grovkillen commented:
Ah!! Forgot that one 🤦‍♂️

Grovkillen commented:
But those two first examples aren't working on my test node. Not with quotes either.

```
On DummyTest#All Do
  LogEntry,"[DummyTest#test1] [DummyTest#test2] [DummyTest#test3] [DummyTest#test4]"
EndOn
```

TD-er committed:
[JSON] Fix a bug in generated JSON, not wrapping '00' as a string

As reported here: https://github.com/letscontrolit/ESPEasy/pull/3439#issuecomment-759269706

TD-er committed:
[Build] Reduce build size for minimal OTA builds

TD-er committed:
[Logs] Format task values in logs with set nr of decimals

As reported here: https://github.com/letscontrolit/ESPEasy/pull/3439#issuecomment-758908074 by @tonhuisman

TD-er commented:
Fixed the log bug as reported by @tonhuisman 
Fixed the issue reported by @Grovkillen which appeared to be not a rules parsing issue, but a JSON formatting issue.

When creating JSON (for the logs for example) the string `00` was not wrapped in quotes as the check to see if something was a numerical was broken in this PR.
So that's a very nice find, which I would not have thought about to test :)

tonhuisman commented:
I also saw a build-issue in Travis about `MDNS.update()` on the ESP32 builds, don't think you tackled that yet?

TD-er commented:
> I also saw a build-issue in Travis about `MDNS.update()` on the ESP32 builds, don't think you tackled that yet?

Working on it, as it is a bigger issue than first expected it to be.

tonhuisman commented:
> as it is a bigger issue than first expected it to be.

IMHO, the biggest issue is that the interface is not the same for ESP8266 and ESP32, but that is also the case for some other interfaces, unfortunately 😞

TD-er committed:
[Cleanup] Fix MDNS builds on ESP32 and reduce minimal OTA build size

TD-er committed:
[ESP32] Fix build issue due to missing include

Missed it in the last commit.....
Need coffee....

TD-er commented:
New test build:
- [ESPEasy_ESP82xx_mega-20201227-130-PR_3439.zip](https://www.dropbox.com/s/9od2vooepfvbrnp/ESPEasy_ESP82xx_mega-20201227-130-PR_3439.zip?dl=0)
- [ESPEasy_ESP32_mega-20201227-130-PR_3439.zip](https://www.dropbox.com/s/6b1yvl9hre10kpd/ESPEasy_ESP32_mega-20201227-130-PR_3439.zip?dl=0)

TD-er committed:
[Build] ESP32 factory bin files are 64k larger so don't reject those

TD-er committed:
[Build] Rename test_ESP32_4M316k_lolin_d32_pro to match build script checks

The copy script matches `4M316k` to validate build size.
However the old env name did not contain this and thus was the file always copied to "rejected" folder

TD-er commented:
If there's nothing else showing up from tests, I would like to merge it  tonight and make a new nighly build.
If I continue here, there will be a bigger chance of causing merge conflicts with other PRs as this one already has become way bigger than originally expected.

tonhuisman commented:
I'll only be available for testing until later tonight (currently wfh, but in on-line meetings, so some time to burn 😎)

TD-er commented:
Well I can also merge it tomorrow morning :)
It is not like we have a deadline.

tonhuisman commented:
This last build 130 (`normal`) is running fine on my ESP8266 test-unit.
Just fired up an ESP32 and installed the `test` build on it, and the first thing after flashing I saw was this serial log:
```
161 : Error  : System: Error - Too many Plugins. PLUGIN_MAX = 85
394 : Info   : INFO : Plugins: 85 [Normal] [TEST ESP32] (ESP32 SDK v3.2.3-14-gd3e562907)
395 : Error  : Programming error! - Increase PLUGIN_MAX
```
And when further inspecting the log this was what I saw:
```
72037 : Info   : EVENT: Clock#Time=Wed,21:12
94575 : Info   : WD   : Uptime 2 ConnectFailures 1 FreeMem 202828 WiFiStatus 3 ESPeasy internal wifi status: Conn. IP Init
dhcps: send_nak>>udp_sendto result 0
```
And all has stopped ⚠️ (web interface dead as well) so my guess is this has to do with the mDNS changes.
RSSI is -67 dB (Accesspoint, Netgear WNDR4300, is about 3 meter away, with free line of sight to the unit)

After a couple of minutes, this log line was added:
```
dhcps: send_nak>>udp_sendto result 0
```
Then I reset the unit, and it all seems to be working again, could be an initialization issue for mDNS (hadn't seen that module activated/enabled before)

You might need to make a few minor adjustments 🤔

TD-er commented:
> You might need to make a few minor adjustments 🤔

Have you seen the rest of this PR?
Those were all the result of some "quick fixes" and "minor tweaks" and started with "oh that's a simple request, will make a quick fix".

TD-er committed:
[ESP32] Increase DEVICES_MAX for ESP32 builds (> 85 plugins in build)

TD-er commented:
Made a new test build, with the max devices increased.
- [ESPEasy_ESP82xx_mega-20201227-133-PR_3439.zip](https://www.dropbox.com/s/kh6jpbzl2vl8nnn/ESPEasy_ESP82xx_mega-20201227-133-PR_3439.zip?dl=0)
- [ESPEasy_ESP32_mega-20201227-133-PR_3439.zip](https://www.dropbox.com/s/e9gv9sonl1k8l7c/ESPEasy_ESP32_mega-20201227-133-PR_3439.zip?dl=0)
@tonhuisman Could you test ESP32 build to see if it is now more responsive?

tonhuisman commented:
This build is indeed more stable on the ESP32, thanks 👍 
And the mDNS is working, though I see some restarts of that service/plugin. Don't know why that happens, Wifi is stable (RSSI -45 dB to a Deco M9 Plus at ca. 1.5 meter, but it is like the MQTTImport (piggybacking on Domoticz MQTT), that also reconnects quite often. May be the Mosquitto service on my RPi is acting up or disconnecting often, but I haven't watched that closely yet.
I'd say, don't change a thing, it's probably best to merge all these fixes now.

TD-er commented:
> I'd say, don't change a thing, it's probably best to merge all these fixes now.

If it is broken now, it's your fault ;)

AloisKlingler commented:
sorry for hijacking this for a question:

TimerSet,3,(abs([VAR#9])+1)
does not work (it's not calculated). I always have to calc the abs([VAR#9]) in a step before and assign it.
I also tried TimerSet,3,{abs([VAR#9])+1} - doesn't work too.
What am I doing wrong?

also:
if ([Z3#SR]-[VAR#8]) > ([Z3#SL]-[VAR#8])
doesn't work, rules processing stops there (the line before is the last processed and logged one).
I also tried if {[Z3#SR]-[VAR#8]} > {[Z3#SL]-[VAR#8]} - doesn't work too.
Do I just have here a wrong brackets?

Thank you! :-)

BR
Alois

TD-er commented:
When using a computation, you must use the `let` command as that one for sure does call the calculate function.
The main problem I still have to tackle is that the rules handling first does perform the replacements and then the calculations.
So if the outer operations are the replacement ones (`{...}`) they will not be called after the calculation has been done.
Something similar may apply to the non working if statement.
It is best to assign the computation to a variable first and compare those.

AloisKlingler commented:
Thank you, clear and fast. 🥇 :-)

Br
Alois
